mne_bids/write.py

```diff
@@ -2158,3 +2163,400 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
     logger.info(f'Writing crosstalk file to {out_path}')
     out_path.mkdir()
     shutil.copyfile(src=fname, dst=str(out_path))
+
+
+def _get_daysback(
+    *,
+    bids_path: BIDSPath,
+    rng: np.random.Generator
+) -> int:
+    """Try to find a suitable "daysback" for anonymization."""
+    raw = read_raw_bids(bids_path=bids_path, verbose='error')
+    daysback_min, daysback_max = get_anonymization_daysback([raw])
+    # We only estimated the range based on one file, so let's add a little
+    # slack here …
+    daysback_min += 356
+    daysback_max -= 365
+
+    # Pick one randomly
+    daysback = int(rng.choice(list(range(daysback_min, daysback_max + 1))))
+
+    return daysback
+
+
+def _check_crosstalk_path(bids_path: BIDSPath) -> bool:
+    is_crosstalk_path = (
+        bids_path.datatype == 'meg' and
+        bids_path.suffix == 'meg' and
+        bids_path.acquisition == 'crosstalk' and
+        bids_path.extension == '.fif'
+    )
+    return is_crosstalk_path
+
+
+def _check_finecal_path(bids_path: BIDSPath) -> bool:
+    is_finecal_path = (
+        bids_path.datatype == 'meg' and
+        bids_path.suffix == 'meg' and
+        bids_path.acquisition == 'calibration' and
+        bids_path.extension == '.dat'
+    )
+    return is_finecal_path
+
+
+@verbose
+def anonymize_dataset(bids_root_in, bids_root_out, daysback='auto',
+                      subject_mapping='auto', datatypes=None,
+                      random_state=None, verbose=None):
+    """Anonymize a BIDS dataset.
+
+    Parameters
+    ----------
+    bids_root_in : path-like
+        The root directory of the input BIDS dataset.
+    bids_root_out : path-like
+        The directory to place the anonymized dataset into.
+    daysback : int | 'auto'
+        Number of days by which to move back the recording date in time. If
+        ``'auto'``, tries to randomly pick a suitable number.
+    subject_mapping : dict | callable | 'auto' | None
+        How to anonymize subject IDs. If a dictionary, maps the original IDs
+        (keys) to the anonymized IDs (values). If a function, must be one that
+        accepts the original IDs as a list of strings and returns a dictionary
+        with original IDs as keys and anonymized IDs as values. If ``'auto'``,
+        automatically produces a mapping (zero-padded numerical IDs) and prints
+        it on the screen. If ``None``, subject IDs are not changed.
+    datatypes : list of str | str | None
+        Which data type to anonymize. If can be ``meg``, ``eeg``, ``ieeg``, or
+        ``anat``. Multiple data types may be passed as a collection of strings.
+        If ``None``, try to anonymize the entire input dataset.
+    %(random_state)s
+        The RNG will be used to derive ``daysback`` and ``subject_mapping`` if
+        they are ``'auto'``.
+    %(verbose)s
+    """
+    from mne_bids import update_sidecar_json  # avoid circular import
```

Comment:
should go into a utils file?

Reply:
What do you mean exactly?

Also I despise `utils` modules, they tend to become a catch-all where one dumps everything

Reply:
1. ramensoup() and noodles() is in file soup.py and friednoodles() and pepper() is in file maincourse.py
2. ramensoup() and friednoodles() both call pepper() and noodles()

Circular import, right? Solution -> move pepper() and noodles() into pantry.py. Basically circular import indicates pathological organization that needs to be fixed (in most cases ...)

Reply:
I'd argue that the entire Python import system is broken, at least when comparing it to modern systems like the one in TypeScript…

I can try to see if a relative import works. Why aren't we using those anyway, @sappelhoff?

Reply:
@jasmainak I fixed it by importing from `mne_bids.sidecar_updates`: c8ad05a

Seems one of our issues is that we import ~everything into the `mne_bids` namespace …

Reply:
> I can try to see if a relative import works. Why aren't we using those anyway, @sappelhoff?

I find relative imports unreadable and confusing. Also, PEP8 says:

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path)

https://www.python.org/dev/peps/pep-0008/

Reply:
yeah, this looks like the right solution. Don't import everything into `mne_bids` namespace